helix-view/src/document.rs

```diff
@@ -614,11 +616,37 @@ impl Document {
 
     /// Select text within the [`Document`].
     pub fn set_selection(&mut self, view_id: ViewId, selection: Selection) {
+        // clear selection history created by [`push_selection`]
+        self.selections_history
+            .entry(view_id)
+            .and_modify(|hist| hist.clear());
```

Comment:
Rather than modifying `set_selection` and wiping everything if the selection is moved, I'd do a check in `shrink_selection` if the selection on the stack is a sub-range of the current range / overlaps the current range. If yes, use it, if no, wipe the stack there. That way there's no extra logic happening any time `set_selection` is called.